cadc-datalink/src/main/java/org/opencadc/datalink/ServiceDescriptorTemplate.java

```diff
@@ -0,0 +1,204 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.log4j.Logger;
+
+/**
+ * Datalink Service Descriptor Template
+ */
+public class ServiceDescriptorTemplate {
+    private static final Logger log = Logger.getLogger(ServiceDescriptorTemplate.class);
+
+    // The descriptor name, must contain only letters, numbers, or a dash.
+    private final String name;
+
+    // VOTable describing the descriptor.
+    private final String template;
+
+    // List of ID attributes from the VOTABLE INFO elements in the template.
+    private final List<String> identifiers;
+
+    public ServiceDescriptorTemplate(final String name, final String template) {
+        if (!StringUtil.hasLength(name)) {
+            throw new IllegalArgumentException("name cannot be null or empty.");
+        }
+        if (!StringUtil.hasLength(template)) {
+            throw new IllegalArgumentException("template cannot be null or empty.");
+        }
+        if (!isValidString(name)) {
+            throw new IllegalArgumentException("Invalid descriptor name: " + name);
+        }
+
+        this.name = name;
+        this.template = template;
+        this.identifiers = parseIdentifiers(template);
+    }
+
+    /**
+     * Get the descriptor name.
+     * @return the descriptor name.
+     */
+    public String getName() {
+        return this.name;
+    }
+
+    /**
+     * Get the descriptor template.
+     * @return the descriptor template.
+     */
+    public String getTemplate() {
+        return this.template;
+    }
+
+    /**
+     * Get the list of identifiers from the template.
+     * The identifiers are the ID attributes from the VOTABLE INFO elements.
+     *
+     * @return the list of identifiers.
+     */
+    public List<String> getIdentifiers() {
+        return this.identifiers;
+    }
+
+    /**
+     * Validate that a string is not be null or empty, and contains only letters, numbers, or a dash.
+     *
+     * @param input the string to validate.
+     */
+    private boolean isValidString(String input) {
+        if (!StringUtil.hasLength(input)) {
+            return false;
+        }
+        return input.matches("[a-zA-Z0-9-]+");
+    }
+
+    /**
+     * Parse the template and extract the identifiers. A template must have at least one
+     * INFO element with an ID attribute in the document root. The template must also have
+     * a group element named 'inputParams' containing a PARAM element with a ref attribute
+     * that matches the ID attribute of the INFO element.
+     */
+    private List<String> parseIdentifiers(String template) {
+        VOTableReader reader = new VOTableReader();
+        VOTableDocument votable;
+        try {
+            votable = reader.read(template);
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Error reading VOTable template: " + e.getMessage());
+        }
+
+        // List of ID's from the INFO elements
```

Comment:
This and the check below (Validate that all ID's have a corresponding ref) are useless because the XML parser already checked that and threw the exception from read(). As noted in the test, you already confirmed this to be the case from the message always being that "ID/IDREF binding" thing

Reply:
Removed redundant checks in the class.